### PR TITLE
overlord/snapstate: apply flags (devmode) on revert

### PR DESCRIPTION
**Pull request by stolowski**
Apply flags on revert to restore apparmor profile in complain mode if devmode was in use for given snap (fixes https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1620560).

### CODE REVIEW

**Comment by niemeyer** [comment]
That doesn't look like what we agreed to.  The --devmode flag should be restored if the _reverted to revision_ was in devmode back then. This is apparently just copying over whatever flags the update was made with, which means the revert is still not an actual revert.

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -686,6 +686,8 @@ func Revert(s *state.State, name string) (*state.TaskSet, error) {
 	return RevertToRevision(s, name, pi.Revision)
 }
 
+// RevertToRevision returns a set of tasks for reverting to the given version of the snap.
```
Comment (chipaca) [inline overlord/snapstate/snapstate.go]: given *revision* ;-)

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -706,6 +708,7 @@ func RevertToRevision(s *state.State, name string, rev snap.Revision) (*state.Ta
 	}
 	ss := &SnapSetup{
 		SideInfo: snapst.Sequence[i],
+		Flags:    SnapSetupFlags(snapst.Flags),
```
Comment (chipaca) [inline overlord/snapstate/snapstate.go]: hmmm... what happens if you try to revert a snap that is in trymode?

**Comment by chipaca** [comment]
Gustavo, there is no "devmode of revision". We'd have to migrate to a state that stored flags per revision. Also it feels weird :-)

**Comment by stolowski** [comment]
@niemeyer yes, I was aware of that, but that's the discussion I had earlier with John:

 John Lenton, [07.09.16 10:55]
@pstolowski in snapstate.RevertToRevision, add the flags from current to the snapsetup

Pawel Stolowski, [07.09.16 10:56]
@Chipaca yes, I thought about that. but that means we apply current flags to old revision, which may not be really "revert" if old revision didn't use devmode?

John Lenton, [07.09.16 10:57]
@pstolowski devmode or not is a snap property, not a revision property

**Comment by niemeyer** [comment]
> Gustavo, there is no "devmode of revision". We'd have to migrate to a state that stored flags per revision. Also it feels weird :-)

I don't understand the feeling of it being weird. Can you provide some insight into it? For example, revision 12 is in devmode, I install revision 42 without it.. breaks. Revert... I would expect to get an actual reversion, to what it was. Not to what I moved towards.

**Comment by niemeyer** [comment]
Perhaps the simplest and less awkward to do right now would be to force the choice, adding --devmode and --jailmode to the revert command.

**Comment by chipaca** [comment]
I've been thinking about why I find it weird, and it's possibly me resisting the very invasive change that would be making flags be per revision. However all the surprising scenarios I can think of involve devmode, where you need to explicitly refresh (and explicitly change the flags) for a revert to even be possible. That is: if you have something installed in jailmode, then a generic `snap refresh` will update your thing conserving that flag, and `snap revert` should also conserve that flag, this is very clear. Whether a `snap refresh --devmode foo` should preserve that devmode on revert is less clear to me, given that `snap refresh foo` would drop that flag.

So perhaps, we want to add the flags to revert, and carry jailmode but not devmode over?

**Pull request closed without merging**